src/main/kotlin/org/jitsi/jicofo/conference/source/ValidatingConferenceSourceMap.kt

```diff
@@ -18,67 +18,302 @@
+
+        /**
+         * Checks if a give [EndpointSourceSet] is internally valid. The checks include:
+         * -- FID groups have exactly 2 SSRCs
+         * -- Grouped sources have an MSID
+         * -- Sources within a group have the same MSID
+         * -- There are no MSID conflicts
+         */
+        @Throws(ValidationFailedException::class)
+        private fun validateEndpointSourceSet(endpointSourceSet: EndpointSourceSet) {
+            endpointSourceSet.ssrcGroups.forEach { group ->
+                if (group.ssrcs.isEmpty()) {
+                    throw IllegalStateException("Empty group should have been filtered out.")
+                }
+                var groupMsid: String? = null
+                group.ssrcs.forEach { ssrc ->
+                    // We are doing a linear search in a list, but the list only has sources for one endpoint, so
+                    // its size is limited to [maxSsrcsPerUser].
+                    val source = endpointSourceSet.sources.find { it.ssrc == ssrc }
+                        ?: throw IllegalStateException(
+                            "Groups with SSRCs that have no corresponding source should have been filtered out."
+                        )
+
+                    if (group.semantics == SsrcGroupSemantics.Fid && group.ssrcs.size != 2) {
+                        throw InvalidFidGroupException(group.ssrcs)
+                    }
+
+                    if (source.msid == null) {
+                        throw RequiredParameterMissingException("msid")
+                    }
+                    if (groupMsid == null) {
+                        groupMsid = source.msid
+                    }
+
+                    if (source.msid != groupMsid) {
+                        throw GroupMsidMismatchException(group.ssrcs)
+                    }
+                }
+            }
+
+            // This groups all sources according to the extended group that they belong to, i.e. a SIM group gets
+            // associated with all of its sources plus the sources in FID groups associated with the SIM group's
+            // sources. For example given these groups:
+            // SIM(1, 2, 3)
+            // FID(1, 4)
+            // FID(2, 5)
+            // FID(3, 6)
+            // FID(111, 222)
+            // And non-grouped source 333 this will associate the SIM group with sources 1..6, the lonely FID group
+            // with sources 111 and 222, and a dummy single-source group with source 333.
+            //
+            // This is technically O(s*g^2) where s is the number of sources and g is the number of groups. But in
+            // practice both s and g are very small (at most 20 and 4 respectively).
+            val grouped: Map<SsrcGroup, List<Source>> =
+                endpointSourceSet.sources.groupBy { groupBySimulcastGroup(it.ssrc, endpointSourceSet.ssrcGroups) }
+
+            // Here we check for MSID conflicts, i.e. we make sure that each group has a unique MSID.
```

Comment:
Will this make it harder in the future to support endpoints with multiple video sources?  I'm thinking that the test should be that if the msid is the same, then the epId should be the same.  But maybe this should just get a comment for now, since we don't support that case yet.

Reply:
I don't know. Wouldn't the client use a new MSID for its other stream? Honestly, I don't understand why jicofo does any validation apart from detecting conflicts between endpoints, but I'm hesitant to remove it (this check is effectively there in the existing code, but it's not clear if the intention is to check for colisions within the same endpoint or not, since it works on the map of all sources in the conference).

Reply:
I _think_ clients use the same MSID for all sources in a PeerConnection?  But actually I'm not sure.  Conceptually I think it has to do with whether they're supposed to be synchronized, but I'm not 100% positive what browsers actually do.

Reply:
I think I agree. I don't see a reason why jicofo would enforce this. Do you mind leaving it for a follow-up PR?

Reply:
No, that's fine.